contracts/deploy/02-inbox/02-arb-to-gnosis-inbox.ts

```diff
@@ -1,56 +1,78 @@
 import { HardhatRuntimeEnvironment } from "hardhat/types";
 import { DeployFunction } from "hardhat-deploy/types";
 import getContractAddress from "../../deploy-helpers/getContractAddress";
+import { ethers } from "hardhat";
 
 enum SenderChains {
-  ARBITRUM = 42161,
-  ARBITRUM_SEPOLIA = 421614,
+  GNOSIS_MAINNET = 100,
+  GNOSIS_CHIADO = 10200,
   HARDHAT = 31337,
 }
+
 const paramsByChainId = {
-  ARBITRUM: {
-    epochPeriod: 3600, // 1 hours
-  },
-  ARBITRUM_SEPOLIA: {
-    epochPeriod: 3600, // 1 hours
-  },
+  GNOSIS_MAINNET: {},
+  GNOSIS_CHIADO: {},
   HARDHAT: {
-    epochPeriod: 600, // 10 minutes
+    epochPeriod: 600, // 10 min
+    routerAddress: ethers.constants.AddressZero,
   },
 };
 
-// TODO: use deterministic deployments
 const deployInbox: DeployFunction = async (hre: HardhatRuntimeEnvironment) => {
   const { ethers, deployments, getNamedAccounts, getChainId, config } = hre;
   const { deploy } = deployments;
-  const chainId = Number(await getChainId());
-  const { providers } = ethers;
 
-  const deployer = (await getNamedAccounts()).deployer;
-  console.log("deployer: %s", deployer);
+  // fallback to hardhat node signers on local network
+  const deployer = (await getNamedAccounts()).deployer ?? (await hre.ethers.getSigners())[0].address;
+  const chainId = Number(await getChainId());
 
   const { epochPeriod } = paramsByChainId[SenderChains[chainId]];
 
-  const routerNetworks = {
-    ARBITRUM: config.networks.mainnet,
-    ARBITRUM_SEPOLIA: config.networks.sepolia,
-    HARDHAT: config.networks.localhost,
-  };
+  // Hack to predict the deployment address on the sender chain.
+  // TODO: use deterministic deployments
 
   // ----------------------------------------------------------------------------------------------
+  const hardhatDeployer = async () => {
+    let nonce = await ethers.provider.getTransactionCount(deployer);
+
+    const arbitrumBridgeAddress = getContractAddress(deployer, nonce + 5);
 
-  const routerChainProvider = new providers.JsonRpcProvider(routerNetworks[SenderChains[chainId]].url);
-  let nonceRouter = await routerChainProvider.getTransactionCount(deployer);
+    const arbSysMock = await deploy("ArbSysMock", {
```

Comment:
There are multiple async calls from here to the end of the function that seems to be independent to each other and can be run concurrently with `Promise.all` for improved performance.

Reply:
already answered

Reply:
Thanks! In general, I consider async operations that runs concurrently should do so. JavaScript made it very easy with the Promise syntax and helps to avoid death by a thousand cuts when code is serialized everywhere.

I do not think it is a mandatory refactor but just wanted to raise the discussion to find some common practices for our systems.

Reply:
I tried deploying contracts concurrently ,with promise.all , I was met with a nonce too low or too high issue, which can be solved by using different deployer wallet for different contracts, that would introduce more complexity, not to mention the scripts uses nonce value to deterministically calculate what will be the address of a future deployment, I believe the trade-off isn't worth it , thoughts?

Reply:
I didn't know of that particular issue! Thanks for bringing it up, all good on serial deployments then :+1: